sbnanalysis/ana/SBNOsc/NueSelection.cxx

```diff
@@ -2,60 +2,345 @@
+
+
+
   // Load configuration parameters
+  fEnergyThreshold =0.;
+  fVertexAssnDis=0.;
+  fNuCount=0;
   fTruthTag = { "generator" };
+  fTrackTag = { "mcreco" };
+  fShowerTag = { "mcreco" };
 
   if (config) {
+    fEnergyThreshold = (*config)["SBNOsc"].get("energy_threshold",123.0).asDouble();
+    fVertexAssnDis = (*config)["SBNOsc"].get("another",1.0).asDouble();
+    //fEnergyThreshold = { (*config)["SBNOsc"].get("energy_threshold",12.34).asDouble() };
     fTruthTag = { (*config)["SBNOsc"].get("MCTruthTag", "generator").asString() };
+    fTrackTag = { (*config)["SBNOsc"].get("MCTrackTag", "mcreco").asString() };
+    fShowerTag = { (*config)["SBNOsc"].get("MCShowerTag","mcreco").asString() };
   }
+  AddBranch("energy_threshold",&fEnergyThreshold);
+  AddBranch("nucount",&fNuCount);
+  AddBranch("vertex_association_distance_uppper_bound",&fVertexAssnDis);
+
 
   hello();
 }
 
 
-void NueSelection::Finalize() {}
+void NueSelection::Finalize() {
+  fOutputFile->cd();
+  fDiffLength->Write();
+  fTrackLength->Write();
+  fGenNueHist->Write();
+  fGenNueFidVolHist->Write();
+  fSelectedNuHist->Write();
+  fShowerEnergy->Write();
+  fEnergeticShowerHist->Write();
+  fVisibleVertexNuEHist->Write();
+  fCGSelectionHist->Write();
+  fRecoSelectionHist->Write();
+  fShowerCutSelectionHist->Write();
+  fSelectedTrueNue->Write();
+  fNodEdxNuHist->Write();
+
+  fMuShowerSelectedNu->Write();
+  fEShowerSelectedNu->Write();
+  fGammaShowerSelectedNu->Write();
+  fOtherShowerSelectedNu->Write();
+
+  fShowerdEdx->Write();
+  fEShowerdEdx->Write();
+  fGammaShowerdEdx->Write();
+  fOtherShowerdEdx->Write();
+  fMuShowerdEdx->Write();
+  fPositronShowerdEdx->Write();
+
+  fGenHist->Write();
+  fGenBarNueHist->Write();
+  fGenNumuHist->Write();
+  fGenOtherHist->Write();
+}
+
 
 
 bool NueSelection::ProcessEvent(const gallery::Event& ev, std::vector<Event::Interaction>& reco) {
+
   if (fEventCounter % 10 == 0) {
     std::cout << "NueSelection: Processing event " << fEventCounter << " "
               << "(" << fNuCount << " neutrinos selected)"
               << std::endl;
   }
   fEventCounter++;
 
-  // Grab a data product from the event
+  // Grab data products from the event
   auto const& mctruths = \
     *ev.getValidHandle<std::vector<simb::MCTruth> >(fTruthTag);
+  auto const& mctracks = \
+    *ev.getValidHandle<std::vector<sim::MCTrack> >(fTrackTag);
+  auto const& mcshowers = \
+    *ev.getValidHandle<std::vector<sim::MCShower> >(fShowerTag);
+
+ //Conversion gap cut: mark visibility of vertices
+ std::vector<bool> IsVisibleVertex;
+  for (size_t i=0;i<mctruths.size();i++) {
+    auto const& mctruth = mctruths.at(i);
+    auto const& nu = mctruth.GetNeutrino();
+    // condition 1: nu vertex within active volume
+    double vx = nu.Nu().Vx();
+    double vy = nu.Nu().Vy();
+    double vz = nu.Nu().Vz();
+    bool InActiveVol = (((-199.15 < vx && vx < -2.65) || (2.65 < vx && vx < 199.15)) && (-200 < vy && vy < 200) && (0 < vz && vz < 500));
```

Comment:
This'll depend on the detector geometry, but @gputnam is adding some utilities for grabbing the fiducial volume for each detector in #8, so maybe can switch to using that once it's available.